File: src/core/Ledger.py
import core.Account
import core.AccountType
import core.AccountingException
import core.Agent
import core.Bank
import core.BaselWeighting
import core.Loan
import core.Transaction


class LedgerType:
    LOAN = 1
    CAPITAL = 2
    DEPOSIT = 3
    CASH = 4


class Ledger:
    def __init__(self, string: str, string2: str, accountType: AccountType, ledgerType: LedgerType) -> None:
        self.name: str = string2
        self.type: AccountType = accountType
        self.ledgertype: LedgerType = ledgerType
        self.transactions: List[Transaction] = []
        self.accounts: Dict[int, Account] = {}

        self.debug = False
        self.postTransactions = True
        self.long_balance = 0
        self.frozen = False
        self.recordTransactions = False
        self.accountId = -1
        self.turnover = 0
        self.lastTotalDeposits = 0
        self.changed = True

    def getAccountNo() -> int:
        return self.accountId if self.frozen else -1

    def debit(transaction: Transaction, account: Account = None, loan: Loan = None, l: int = None) -> None:
        if loan is not None:
            if self.debug:
                print("debit ledger: " + self.name + " " + loan)
                if account not in self.accounts:
                    self.audit()
                    raise RuntimeException("Attempt to debit account not in ledger" + account)

            if (self.ledgertype == LedgerType.LOAN and self.type == AccountType.ASSET):
                account.addCapitalLoan(loan)
            else:
                account.addLoan(loan)

            self.addTransaction(transaction)
            self.turnover += loan.getCapitalOutstanding()
        elif l is not None and account is not None:
            if self.debug:
                print("debit ledger: " + self.name + " " + l)
                if account not in self.accounts:
                    self.audit()
                    raise RuntimeException("Attempt to debit account not in ledger " + self.name + " " + account)
            l2: int = l * -1 * self.type.polarity()
            if account.deposit + l2 < 0:
                raise RuntimeException("Negative Balance in Account after debit " + account)
            account.deposit += l2
            self.addTransaction(transaction)
            if self.type.polarity() < 0:
                self.turnover += l
        elif l is not None:
            if (not self.frozen) or len(self.accounts) != 1:
                raise RuntimeException("Cannot use with multi-account ledger")
            self.debit(transaction, self.getAccount(), l)
            if self.type.polarity() < 0:
                self.turnover += l
        else:
            raise Exception("not implemented")

    def credit(transaction: Transaction, account: Account = None, loan: Loan = None, l: int = None) -> None:
        if loan is not None and account is not None:
            if self.debug:
                print("credit ledger: " + self.name + " " + loan)
                if account not in self.accounts:
                    raise RuntimeException("Attempt to debit account not in ledger" + account)

            if self.ledgertype == LedgerType.LOAN and self.type == AccountType.ASSET:
                account.addCapitalLoan(loan)
            else:
                account.addLoan(loan)

            self.addTransaction(transaction)
            self.turnover += loan.getCapitalOutstanding()
        elif loan is not None and l is not None:
            if loan.writeOff(l):
                loan.remove()
            self.addTransaction(transaction)
        elif l is not None and account is not None:
            if self.debug:
                self.audit()
                print("credit ledger = " + self.name + " : " + l + " from " + account.owner.name)
                if account not in self.accounts:
                    self.audit()
                    raise RuntimeException("Attempt to credit account not in ledger " + self.name + ": " + account)
            l2: int = l * self.type.polarity()
            if (account.deposit + l2 < 0):
                raise RuntimeException("Negative Balance in Account after credit " + account)
            account.deposit += l2
            self.addTransaction(transaction)
            if self.type.polarity() > 0:
                self.turnover += l
        elif l is not None:
            if (not self.frozen) or self.accounts.size() != 1:
                raise RuntimeException("Cannot use with multi-account ledger")
            self.credit(self.getAccount(), l, transaction)
            if self.type.polarity() > 0:
                self.turnover += l
        else:
            raise Exception("not implemented")

    def payLoan(loan: Loan, arrl: List[int], transaction: Transaction) -> None:
        loan.makePayment(arrl)
        self.addTransaction(transaction)

    def addAccount(account: Account) -> None:
        if account.deposit != 0:
            raise AccountingException("Account deposit != 0 " + account.getId())
        if account.getId() in self.accounts:
            raise AccountingException("Ledger already contains account" + account.getId())
        if self.frozen:
            raise AccountingException("Attempt to add account to frozen ledger" + account.getId())
        if account.getID() in self.accounts:
            raise AccountingException("Duplicate account key in ledger: " + self.name)
        self.accounts.put(account.getId(), account)
        account.ledger = self.name

    def addAccountAndClose(account: Account) -> None:
        if account.deposit != 0:
            raise AccountingException("Account deposit != 0 " + account.getId())
        self.addAccount(account)
        self.frozen = True
        if len(self.accounts) == 1:
            self.accountId = account.accountId

    def removeAccount(account: Account) -> None:
        try:
            self.accounts.remove(account.accountId)
        except Exception as e:
            raise RuntimeException(e)

    def getAccount() -> Account:
        if len(self.accounts) != 1:
            raise RuntimeException("Requested single account from multi-account ledger:" + self.name)
        return self.accounts.values()[0]

    def getAccount(n: int) -> Account:
        return self.accounts.get(n)

    def debitPolarity() -> int:
        return self.type.polarity() * -1

    def creditPolarity() -> int:
        return self.type.polarity()

    def getType() -> AccountType:
        return self.type

    def getLedgerType() -> LedgerType:
        return self.ledgertype

    def containsLoan(loan: Loan) -> bool:
        return self.getAccount().holdsLoan(loan)

    def audit() -> None:
        print("Audit : " + self.name)
        for account in self.accounts.values():
            System.out.println("\t" + account.accountId + " " + account.bank.name)

    def total() -> int:
        if self.ledgertype == LOAN:
            if self.type == AccountType.ASSET:
                return self.totalCapital()
            return self.totalLoans()
        elif self.ledgertype == CAPITAL:
            return self.totalCapital()
        elif self.ledgertype == DEPOSIT:
            if self.changed:
                self.lastTotalDeposits = self.totalDeposits()
                self.changed = false
            return self.lastTotalDeposits
        raise RuntimeException("Unhandled entry type in ledger (new enum??)")

    def totalDeposits() -> int:
        if self.accounts is not None:
            return sum(account.getDeposit() for account in self.accounts.values())
        return 0

    def totalCapital() -> int:
        if self.accounts is not None:
            if self.type == AccountType.EQUITY:
                assert len(self.accounts) == 1
                account = self.accounts.values()[0]
                return account.getTotalCapital() + account.getDeposit()
            else:
                return sum(account.getTotalCapital() for account in self.accounts.values())
        return 0

    def recalculateVariableLoans(d: float) -> None:
        if self.accounts is None:
            return
        for account in self.accounts.values():
            for loan in account.debts.values():
                loan.interestRate = d
                loan.setCompoundSchedules(loan.payIndex)

    def totalLoans() -> int:
        if self.accounts is not None:
            return sum(account.getTotalDebt() for account in self.accounts.values())
        return 0

    def riskWeightedTotalLoans() -> int:
        out: int = 0
        assert self.name != "loan"
        if self.accounts is not None:
            for (Account account : self.accounts.values()) {
                l += sum(loan.getCapitalOutstanding() * BaselWeighting.riskWeighting(loan) for loan in account.capital_loans.values())
        return l

    def addTransaction(transaction: Transaction) -> None:
        self.changed = True
        if self.postTransactions:
            self.transactions.add(transaction)

    def getTurnover() -> int:
        l: int = self.turnover
        self.turnover = 0
        return l

    def getName() -> str:
        return self.name

    def toString() -> str:
        return self.name

    def printAccounts() -> None:
        print("Ledger: " + self.name)
        for account in self.accounts.values():
            print("Deposits")
            print("\t" + account)
            if len(account.debts) > 0:
                print("\nLoans")
                for loan in account.debts.values():
                    print("\t\tDebt:\t" + loan)
            if len(account.capital_loans) <= 0:
                continue
            print("\nCapital")
            for loan in account.capital_loans.values():
                print("\t\tCapital:\t" + loan)
        print()

File: src/core/Ledger.java
/*
 * Decompiled with CFR 0_132.
 */
package core;

import core.Account;
import core.AccountType;
import core.AccountingException;
import core.Agent;
import core.Bank;
import core.BaselWeighting;
import core.LedgerType;
import core.Loan;
import core.Transaction;
import java.io.PrintStream;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;

public class Ledger {
    public String name;
    private boolean debug = false;
    public static boolean postTransactions = true;
    private long balance = 0L;
    private boolean frozen = false;
    private boolean recordTransactions = false;
    public int accountId = -1;
    private AccountType type;
    public LedgerType ledgertype;
    public List<Transaction> transactions;
    public HashMap<Integer, Account> accounts;
    private long turnover = 0L;
    private long lastTotalDeposits;
    boolean changed = true;

    public Ledger(String string, String string2, AccountType accountType, LedgerType ledgerType) {
        this.name = string2;
        this.type = accountType;
        this.ledgertype = ledgerType;
        this.transactions = new ArrayList<Transaction>();
        this.accounts = new HashMap(100);
    }

    public int getAccountNo() {
        if (this.frozen) {
            return this.accountId;
        }
        return -1;
    }

    public void debit(Account account, long l, Transaction transaction) {
        long l2;
        if (this.debug) {
            System.out.println("debit ledger: " + this.name + " " + l);
            if (!this.accounts.containsValue(account)) {
                this.audit();
                throw new RuntimeException("Attempt to debit account not in ledger " + this.name + " " + account);
            }
        }
        if (account.deposit + (l2 = l * (long)(-1 * this.type.polarity())) < 0L) {
            throw new RuntimeException("Negative Balance in Account after debit " + account);
        }
        account.deposit += l2;
        this.addTransaction(transaction);
        if (this.type.polarity() < 0) {
            this.turnover += l;
        }
    }

    public void debit(Account account, Loan loan, Transaction transaction) {
        if (this.debug) {
            System.out.println("debit ledger: " + this.name + " " + loan);
            if (!this.accounts.containsValue(account)) {
                this.audit();
                throw new RuntimeException("Attempt to debit account not in ledger" + account);
            }
        }
        if (this.ledgertype == LedgerType.LOAN && this.type == AccountType.ASSET) {
            account.addCapitalLoan(loan);
        } else {
            account.addLoan(loan);
        }
        this.addTransaction(transaction);
        this.turnover += loan.getCapitalOutstanding();
    }

    public void credit(Account account, Loan loan, Transaction transaction) {
        if (this.debug) {
            System.out.println("credit ledger: " + this.name + " " + loan);
            if (!this.accounts.containsValue(account)) {
                throw new RuntimeException("Attempt to debit account not in ledger" + account);
            }
        }
        if (this.ledgertype == LedgerType.LOAN && this.type == AccountType.ASSET) {
            account.addCapitalLoan(loan);
        } else {
            account.addLoan(loan);
        }
        this.addTransaction(transaction);
        this.turnover += loan.getCapitalOutstanding();
    }

    public void credit(Account account, long l, Transaction transaction) {
        long l2;
        if (this.debug) {
            this.audit();
            System.out.println("credit ledger = " + this.name + " : " + l + " from " + account.owner.name);
            if (!this.accounts.containsValue(account)) {
                this.audit();
                throw new RuntimeException("Attempt to credit account not in ledger " + this.name + ": " + account);
            }
        }
        if (account.deposit + (l2 = l * (long)this.type.polarity()) < 0L) {
            throw new RuntimeException("Negative Balance in Account after credit " + account);
        }
        account.deposit += l2;
        this.addTransaction(transaction);
        if (this.type.polarity() > 0) {
            this.turnover += l;
        }
    }

    public void payLoan(Loan loan, long[] arrl, Transaction transaction) {
        loan.makePayment(arrl);
        this.addTransaction(transaction);
    }

    public void credit(Loan loan, long l, Transaction transaction) {
        if (loan.writeOff(l)) {
            loan.remove();
        }
        this.addTransaction(transaction);
    }

    public void credit(long l, Transaction transaction) {
        if (!this.frozen || this.accounts.size() != 1) {
            throw new RuntimeException("Cannot use with multi-account ledger");
        }
        this.credit(this.getAccount(), l, transaction);
        if (this.type.polarity() > 0) {
            this.turnover += l;
        }
    }

    public void debit(long l, Transaction transaction) {
        if (!this.frozen || this.accounts.size() != 1) {
            throw new RuntimeException("Cannot use with multi-account ledger");
        }
        this.debit(this.getAccount(), l, transaction);
        if (this.type.polarity() < 0) {
            this.turnover += l;
        }
    }

    public void addAccount(Account account) throws AccountingException {
        if (account.deposit != 0L) {
            throw new AccountingException("Account deposit != 0 " + account.getId());
        }
        if (this.accounts.containsKey(account.getId())) {
            throw new AccountingException("Ledger already contains account" + account.getId());
        }
        if (this.frozen) {
            throw new AccountingException("Attempt to add account to frozen ledger" + account.getId());
        }
        if (this.accounts.containsKey(account.getId())) {
            throw new AccountingException("Duplicate account key in ledger: " + this.name);
        }
        this.accounts.put(account.getId(), account);
        account.ledger = this.name;
    }

    public void addAccountAndClose(Account account) throws AccountingException {
        if (account.deposit != 0L) {
            throw new AccountingException("Account deposit != 0 " + account.getId());
        }
        this.addAccount(account);
        this.frozen = true;
        if (this.accounts.size() == 1) {
            this.accountId = account.accountId;
        }
    }

    public void removeAccount(Account account) {
        try {
            this.accounts.remove(account.accountId);
        }
        catch (Exception exception) {
            throw new RuntimeException(exception);
        }
    }

    public Account getAccount() {
        if (this.accounts.size() != 1) {
            throw new RuntimeException("Requested single account from multi-account ledger:" + this.name);
        }
        Object[] arrobject = this.accounts.values().toArray();
        assert (arrobject.length == 1);
        return (Account)arrobject[0];
    }

    public Account getAccount(Integer n) {
        return this.accounts.get(n);
    }

    public int debitPolarity() {
        return this.type.polarity() * -1;
    }

    public int creditPolarity() {
        return this.type.polarity();
    }

    public AccountType getType() {
        return this.type;
    }

    public LedgerType getLedgerType() {
        return this.ledgertype;
    }

    public boolean containsLoan(Loan loan) {
        return this.getAccount().holdsLoan(loan);
    }

    public void audit() {
        System.out.println("Audit : " + this.name);
        for (Account account : this.accounts.values()) {
            System.out.println("\t" + account.accountId + " " + account.bank.name);
        }
    }

    public long total() {
        switch (this.ledgertype) {
            case LOAN: {
                if (this.type == AccountType.ASSET) {
                    return this.totalCapital();
                }
                return this.totalLoans();
            }
            case CAPITAL: {
                return this.totalCapital();
            }
            case CASH: 
            case DEPOSIT: {
                if (this.changed) {
                    this.lastTotalDeposits = this.totalDeposits();
                    this.changed = false;
                }
                return this.lastTotalDeposits;
            }
        }
        throw new RuntimeException("Unhandled entry type in ledger (new enum??)");
    }

    private long totalDeposits() {
        long l = 0L;
        if (this.accounts != null) {
            for (Account account : this.accounts.values()) {
                l += account.getDeposit();
            }
            return l;
        }
        return 0L;
    }

    private long totalCapital() {
        long l = 0L;
        if (this.accounts != null) {
            for (Account account : this.accounts.values()) {
                l += account.getTotalCapital();
                if (this.type != AccountType.EQUITY) continue;
                assert (this.accounts.size() == 1);
                l += account.getDeposit();
            }
        }
        return l;
    }

    public void recalculateVariableLoans(double d) {
        if (this.accounts != null) {
            for (Account account : this.accounts.values()) {
                for (Loan loan : account.debts.values()) {
                    loan.interestRate = d;
                    loan.setCompoundSchedules(loan.payIndex);
                }
            }
        }
    }

    public long totalLoans() {
        long l = 0L;
        if (this.accounts != null) {
            for (Account account : this.accounts.values()) {
                l += account.getTotalDebt();
            }
        }
        return l;
    }

    public long riskWeightedTotalLoans() {
        long l = 0L;
        assert (!this.name.equals("loan"));
        if (this.accounts != null) {
            for (Account account : this.accounts.values()) {
                for (Loan loan : account.capital_loans.values()) {
                    l = (long)((double)l + (double)loan.getCapitalOutstanding() * BaselWeighting.riskWeighting(loan));
                }
            }
        }
        return l;
    }

    private void addTransaction(Transaction transaction) {
        this.changed = true;
        if (postTransactions) {
            this.transactions.add(transaction);
        }
    }

    public long getTurnover() {
        long l = this.turnover;
        this.turnover = 0L;
        return l;
    }

    public String getName() {
        return this.name;
    }

    public String toString() {
        return this.name;
    }

    public void printAccounts() {
        System.out.println("Ledger: " + this.name);
        for (Account account : this.accounts.values()) {
            System.out.println("Deposits");
            System.out.println("\t" + account);
            if (account.debts.size() > 0) {
                System.out.println("\nLoans");
                for (Loan loan : account.debts.values()) {
                    System.out.println("\t\tDebt:\t" + loan);
                }
            }
            if (account.capital_loans.size() <= 0) continue;
            System.out.println("\nCapital");
            for (Loan loan : account.capital_loans.values()) {
                System.out.println("\t\tCapital:\t" + loan);
            }
        }
        System.out.println("\n");
    }

}

